src/interfaces.ts

```diff
@@ -162,3 +167,27 @@ export interface CommentFlag {
     affectedRange: Range;
     codes: number[] | null;
 }
+
+type ValidateHandler = (scope: Scope, files: (BrsFile | XmlFile)[], callables: CallableContainerMap) => void;
+
+export interface CompilerPlugin {
+    name: string;
+    beforePrepublish?: (files: FileObj[]) => void;
+    beforePublish?: (files: FileObj[]) => void;
+    afterPublish?: (files: FileObj[]) => void;
+    programCreated?: (program: Program) => void;
+    programValidateStart?: (program: Program) => void;
+    programValidateEnd?: (program: Program) => void;
+    scopeCreated?: (scope: Scope) => void;
+    scopeDisposed?: (scope: Scope) => void;
+    scopeValidateStart?: ValidateHandler;
+    scopeValidateEnd?: ValidateHandler;
+    fileSourceLoaded?: (source: SourceInfo) => void;
+    fileParsed?: (file: BrsFile) => void;
+    fileValidation?: (file: BrsFile) => void;
+    componentSourceLoaded?: (source: SourceInfo) => void;
+    componentParsed?: (file: XmlFile) => void;
```

Comment:
I don't think we need the `componentSourceLoaded` and `componentParsed` events. The component files should emit the `fileSourceLoaded` and `fileParsed` events instead. We may add additional file types in the future, and the plugin author can detect the type of file in their event handler.